src/storage/query/ScanEdgeProcessor.cpp

```diff
@@ -0,0 +1,173 @@
+    if (kvRet != kvstore::ResultCode::SUCCEEDED) {
+        pushResultCode(to(kvRet), partId_);
+        onFinished();
+        return;
+    }
+
+    std::vector<cpp2::ScanEdge> edgeData;
+    int32_t rowCount = 0;
+    int32_t rowLimit = req.get_limit();
+    int64_t startTime = req.get_start_time(), endTime = req.get_end_time();
+    int32_t blockSize = 0;
+
+    for (; iter->valid() && rowCount < rowLimit && blockSize < FLAGS_max_scan_block_size;
+         iter->next()) {
+        auto key = iter->key();
+        if (!NebulaKeyUtils::isEdge(key)) {
+            continue;
+        }
+
+        EdgeType edgeType = NebulaKeyUtils::getEdgeType(key);
+        if (edgeType < 0) {
+            continue;
+        }
+
+        // only return data within time range [start, end)
+        EdgeVersion version = folly::Endian::big(NebulaKeyUtils::getVersion(key));
+        int64_t ts = std::numeric_limits<int64_t>::max() - version;
+        if (ts < startTime || ts >= endTime) {
+            continue;
+        }
+
+        auto ctxIter = edgeContexts_.find(edgeType);
+        if (ctxIter == edgeContexts_.end()) {
+            continue;
+        }
+
+        auto srcId = NebulaKeyUtils::getSrcId(key);
+        auto dstId = NebulaKeyUtils::getDstId(key);
+        cpp2::ScanEdge data;
+        data.set_src(srcId);
+        data.set_type(edgeType);
+        data.set_dst(dstId);
+        auto value = iter->val();
+        if (returnAllColumns_) {
+            // return all columns
+            data.set_value(value.str());
+        } else if (!ctxIter->second.empty()) {
+            // only return specified columns
+            auto reader = RowReader::getEdgePropReader(schemaMan_, value, spaceId_, edgeType);
+            RowWriter writer;
+            PropsCollector collector(&writer);
+            auto& props = ctxIter->second;
+            collectProps(reader.get(), props, &collector);
+            data.set_value(writer.encode());
+        }
+
+        edgeData.emplace_back(std::move(data));
+        rowCount++;
+        blockSize += key.size() + value.size();
+    }
+
+    resp_.set_edge_schema(std::move(edgeSchema_));
+    resp_.set_edge_data(std::move(edgeData));
+    if (iter->valid()) {
+        resp_.set_has_next(true);
+        resp_.set_next_cursor(std::move(iter->key().str()));
+    } else {
+        resp_.set_has_next(false);
+    }
+    onFinished();
+}
+
+cpp2::ErrorCode ScanEdgeProcessor::checkAndBuildContexts(const cpp2::ScanEdgeRequest& req) {
+    for (const auto& edgeIter : req.get_return_columns()) {
+        int32_t index = 0;
+        EdgeType edgeType = edgeIter.first;
+        std::vector<PropContext> propContexts;
+        auto schema = this->schemaMan_->getEdgeSchema(spaceId_, edgeType);
+        if (!schema) {
+            return cpp2::ErrorCode::E_EDGE_NOT_FOUND;
+        }
+
+        if (returnAllColumns_) {
+            // return all columns
+            edgeContexts_.emplace(edgeType, std::move(propContexts));
```

Comment:
Do we really need to set edgeContexts_ and edgeSchema_ when returning all columns

Reply:
edgeContexts_ and edgeSchema_ are used to store related edgeType, so we can check if we need to scan this edge